NCSU API_Client fundcode dropdown and general form style enhancements

jcraitz opened the pull request:
- adds a fund code drop down to the api_client form
- adds css styles and reworks the form to better use html form standards and the pure-css css framework
- imports some styles from main CORAL pages
- adds in optional commented-out-fields for author and isbn/issn inputs to be added to the description text field
- adds in API flight routes for getting funds for the dropdown
- checks a number of variables for isset or !empty before calling them
- adds in an urgent checkbox for rush orders with a browser native datepicker if the user checks it
- subjective change to make the administering site input (labeled 'Library) to be a single dropdown instead of a multi-select box.

jcraitz commented:
Some screenshots of updated API client form.

![screen shot 2018-01-11 at 5 40 26 pm](https://user-images.githubusercontent.com/3514165/34851232-1aa57930-f6f7-11e7-91f4-53e4f2871806.png)

![screen shot 2018-01-11 at 5 40 55 pm](https://user-images.githubusercontent.com/3514165/34851231-1a9837c0-f6f7-11e7-9421-480dbdb49665.png)

jcraitz commented:
Hi @veggiematts, @t4k, and @PaulPoulain
It's not giving me the option to request a review. Can we still get this in before the freeze?  I'll be available all day Friday, Jan. 12th, for changes, fixes, explanations or discussions.
Cheers!

t4k commented:
I have not worked with the API, but hopefully @veggiematts can look this over.

jcraitz commented:
The basic test plan for this one should just be 
- filling in all the inputs, checking to see if drop downs were populated from the coral db
- submitting (note: there is now validation on a few fields!)
- checking the review page to see if it matches up
- going to the resource in the resource module
- clicking through the different tabs to see if the notes and descriptions were entered correctly

I just realized that I used an Americanized date listing when I added the new urgent date field.  This can be changed if needed.

veggiematts commented:
There's a call to a missing CSS:
*102 open() "/home/coral/www/resources/css/datePicker.css" failed (2: No such file or directory),

I get an error after submitting a resource, instead of the resource summary (the resource is created, however):

> The resource could not be submitted. (error: )

The logs:
```
PHP message: PHP Notice:  Array to string conversion in /home/coral/www/resources/api_client/vendor/mashape/unirest-php/src/Unirest/Request.php on line 405
PHP message: PHP Notice:  Trying to get property of non-object in /home/coral/www/resources/api_client/index.php on line 122" while reading response header from upstream, client: 10.54.190.1, server: coral.local, request: "POST /resources/api_client/index.php HTTP/1.1", upstream: "fastcgi://unix:/run/php/php7.0-fpm.sock:", host: "coral.local", referrer: "http://coral.local/resources/api_client/"


```

veggiematts commented:
Also a few questions:
Why are the fields for author and isbn commented out?
Why is the isbn added to the description field and not to the isbn field directly?

veggiematts commented:
Please also note that a few modifications have to be made in the API to be compatible with #244

jcraitz commented:
- I'll remove the datepicker css. I think that was referencing an earlier version of CORAL's css files.
- The error you're getting is interesting.  For some reason it triggered because there was no resourceID in the body of the response you got back when proposing the resource.  I believe the log is complaining because either there was no $response->body->error or it didn't parse the error correctly.
- I'll investigate.  Do you have error reporting turned on?  I seem to remember that if it's turned on it could foul up the json response from Flight.

I commented out the author and isbn fields because those were local changes requested by our staff and I wasn't sure if the community as a whole would be interested.

The same for the ISBN/ISSN field.  I put it in the description field because it was something that we wanted, but the rest of the community might not.  To put it in the proper field it would require changing more than just the api_client form.  If the community is interested, I can pursue adding the necessary functions in the API if they haven't already been added in recent pull requests.

veggiematts commented:
My error reporting was turned off during testing.
Please note that I get the summary on master.

jcraitz commented:
I've fixed the error listed and can update my pull request, but after rebasing this on to development, I'm getting a new problem.  

It looks like I'll have to change the way I associate the fund code information with the new development changes. No point updating this request until I sort out the new problem.

veggiematts commented:
The new commit in #345 should fix all issues with multiple orders.

However, I found that the administeringSites (Library in the form) were causing the summary error (and prevented further processing after the resource was added). Is that what you've fixed?

Also, in the previous form, Library was a multiple selection list (because you can have several libraries), and it's a one selection dropdown in the new form, why is that?

jcraitz commented:
After updating this one to the changes in development and pulling in #345, I'm till having a few issues.  I'm getting the summary and the resource is creating, but there are a few problems with the resource that's created.  Can't figure out yet if it's all #345's fault, my merging, or something I've done in this request.  It's a lot to juggle all of these PR's at once.

We changed Library to be a single dropdown locally because we didn't have a use case for multiple libraries.  I can remove that part from this pull request if you think others would like to keep that functionality in.

jcraitz commented:
I've figured out the issue. it was the api trying to enter the new workflow from the resource class, but that function was moved to the resourceAcquisition class.  I can update this PR tomorrow, but it will need #345 (with my suggested changes) and #300 to be merged first for any of this to work correctly.

I still need to know what I should do about the changes you had questions about.
1. Should I remove my changes for the administeringsite input?
2. author & ISBN changes
 a. Should I remove these commented out fields?
 b. Should I just uncomment them?
 c. Do I need to fix the ISBN field so that it hooks in to the ISBN table?

veggiematts commented:
> Should I remove my changes for the administeringsite input?

I feel like the multi-select should be kept, because otherwise we're losing a feature.

> author & ISBN changes
> a. Should I remove these commented out fields?
> b. Should I just uncomment them?
>  c. Do I need to fix the ISBN field so that it hooks in to the ISBN table?

I would say that either you keep them commented, or you fix the ISBN field so that it hooks in to the ISBN table, because it would be kind of misleading for the user to have an ISBN field in the form that doesn't go into the ISBN field in Coral, don't you think?
However, at this point, I'm not sure how many people use this form, and how badly is the ISBN feature is needed, so the choice is yours between those two options, I guess.

xsong9 commented:
@jcraitz @veggiematts I'd say hook the ISBN to the ISBN table considering users may use ISBN for de-dup when import new records via csv import function. For author, keep them commented because otherwise it needs to create a new author field. I'm ok with keeping the administeringsite as multi-select.

jcraitz commented:
@veggiematts Unfortunately, I had to force push on a rebase so you'll need to update your copy of this branch for testing.

I've made all the changes discussed.  I've also had to include the changes from #300, so if we merge this, we can close that one.

- kept the author input commented out, but I added a brief description on how to add it back in and how others can use the associated function to add other custom fields to the description text field.
- hooked up the isbn/issn field to the database.
- reverted the administeringSite input to a multi-select

Everything else should be good to go.

veggiematts commented:
Everything is ok, except for the administeringSiteID (Library).
On my instance, it comes like this to the API:
```
    [data:flight\util\Collection:private] => Array
        (
            [user] => API
            [titleText] => resource
            [acquisitionTypeID] => 4
            [administeringSiteID] => Array
            [homeLocationNote] => Stacks
...
```
The value is a string containing the word "Array", instead of being an actual array.
This stops the adding process at this point, leading to an incomplete resource.

jcraitz commented:
@veggiematts I've been trying to get this error for over an hour and can't reproduce it.  Are you sure you pulled down the most recent commit of this and not an older copy?  I did a rebase and might have lost a few things in the merge.

veggiematts commented:
Okay, I finally found where this bug comes from: Unirest.
By switching to v2.6.6, administeringSites are processed as they should:

```
    [data:flight\util\Collection:private] => Array
        (
          ...
            [acquisitionTypeID] => 4
            [administeringSiteID] => Array
                (
                    [0] => 1
                    [1] => 2
                )
...
```

There must be a bug in Unirest v3.0.4.

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing mashape/unirest-php (v3.0.4)
  - Installing mashape/unirest-php (v2.6.6)
    Downloading: 100%

Writing lock file
Generating autoload files
```

This is something we should keep in mind, but I'm ok with merging this PR now.

subratlubu commented:
> Some screenshots of updated API client form.
> 
> ![screen shot 2018-01-11 at 5 40 26 pm](https://user-images.githubusercontent.com/3514165/34851232-1aa57930-f6f7-11e7-91f4-53e4f2871806.png)
> 
> ![screen shot 2018-01-11 at 5 40 55 pm](https://user-images.githubusercontent.com/3514165/34851231-1a9837c0-f6f7-11e7-9421-480dbdb49665.png)

Hi Jason Raitz,
I have installed Coral and run it on a localhost server. I have installed "flight, unirest, league-oauth2-client" under "resources/api" and "resources/api_client" directory through Composer. But I am unable to see the api_client form. The below error screenshot is attached for your reference. Kindly help me to resolve the issue.
![coral_api_client](https://github.com/user-attachments/assets/cc3f6ddc-79c9-4a7b-b41e-c6ba68ca2a2c)
